File: src/main/java/net/nodus/security/auth/service/OAuthLoginService.java
package net.nodus.security.auth.service;

import net.nodus.database.account.OAuthProvider;
import net.nodus.database.account.UserAccount;
import net.nodus.database.account.UserAccountRepository;
import org.springframework.stereotype.Service;

@Service
public class OAuthLoginService {

    private final UserAccountRepository userAccountRepository;
    private final JwtTokenService jwtTokenService;
    private final RefreshTokenService refreshTokenService;

    public OAuthLoginService(
        UserAccountRepository userAccountRepository,
        JwtTokenService jwtTokenService,
        RefreshTokenService refreshTokenService
    ) {
        this.userAccountRepository = userAccountRepository;
        this.jwtTokenService = jwtTokenService;
        this.refreshTokenService = refreshTokenService;
    }

    public OAuthLoginResult loginGoogleUser(String providerId, String email, String name) {
        UserAccount user = userAccountRepository
            .findByProviderAndProviderIdAndDeletedAtIsNull(OAuthProvider.GOOGLE, providerId)
            .orElseGet(() -> userAccountRepository.save(
                new UserAccount(email, name, OAuthProvider.GOOGLE, providerId)
            ));

        RefreshTokenService.IssuedRefreshToken refreshToken = refreshTokenService.issue(user);

        return new OAuthLoginResult(
            jwtTokenService.createAccessToken(user),
            refreshToken.token()
        );
    }

    public record OAuthLoginResult(String accessToken, String refreshToken) {

    }
}

File: src/main/kotlin/net/nodus/auth/service/OAuthLoginService.kt
package net.nodus.auth.service

import net.nodus.account.OAuthProvider
import net.nodus.account.UserAccount
import net.nodus.account.UserAccountRepository
import org.springframework.stereotype.Service

@Service
class OAuthLoginService(
    private val userAccountRepository: UserAccountRepository,
    private val jwtTokenService: JwtTokenService,
    private val refreshTokenService: RefreshTokenService,
) {

    fun loginGoogleUser(providerId: String, email: String, name: String): OAuthLoginResult {
        val user = userAccountRepository.findByProviderAndProviderIdAndDeletedAtIsNull(OAuthProvider.GOOGLE, providerId)
            ?: userAccountRepository.save(
                UserAccount(
                    email = email,
                    name = name,
                    provider = OAuthProvider.GOOGLE,
                    providerId = providerId
                )
            )
        val refreshToken = refreshTokenService.issue(user)

        return OAuthLoginResult(
            accessToken = jwtTokenService.createAccessToken(user),
            refreshToken = refreshToken.token,
        )
    }
}

data class OAuthLoginResult(
    val accessToken: String,
    val refreshToken: String,
)